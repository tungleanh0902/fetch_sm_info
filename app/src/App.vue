<template>
  <div id="app" :class="themeClass">
    <NetworkSelector />
    <AppHeader />
    <div id="app-content">
      <div id="bar-container">
        <CookieBar v-if="!isMobileApp" />
        <MaintenanceBar />
        <DisconnectedBar />
      </div>
      <UserMenu v-if="!isMobileApp" />
      <router-view name="session" />
      <router-view />
    </div>
    <MobileMenu />
  </div>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import AppHeader from "common/AppHeader"
import CookieBar from "common/CookieBar"
import DisconnectedBar from "common/DisconnectedBar"
import UserMenu from "account/UserMenu"
import MaintenanceBar from "common/MaintenanceBar"
import MobileMenu from "common/MobileMenu"
import NetworkSelector from "common/NetworkSelector"
import store from "./vuex/store"

export default {
  name: `app`,
  components: {
    AppHeader,
    CookieBar,
    DisconnectedBar,
    UserMenu,
    MaintenanceBar,
    MobileMenu,
    NetworkSelector,
  },
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`]),
    isMobileApp() {
      return this.session.mobile
    },
    themeClass() {
      return !this.$route.meta.networkSpecificRoute
        ? `lunie-light`
        : this.network
    },
  },
  store,
}
</script>

<style>
@import "./styles/app.css";
</style>
