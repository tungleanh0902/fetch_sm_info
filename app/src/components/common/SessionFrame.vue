<template>
  <transition name="component-fade" mode="out-in">
    <div
      v-focus-last
      class="session-frame"
      tabindex="0"
      @keyup.esc="closeModal()"
    >
      <div class="session-outer-container">
        <div
          v-if="
            (icon.startsWith(`/img`) || icon.startsWith(`https`)) &&
            !isExtension
          "
          class="icon-circle"
        >
          <img class="icon-image" :src="icon" />
        </div>
        <i
          v-else-if="icon.length > 0 && !isExtension"
          class="material-icons notranslate circle modal-icon"
          >{{ icon }}</i
        >
        <div class="session">
          <div class="session-header">
            <a :class="{ invisible: hideBack }" @click="goBack">
              <i class="material-icons notranslate circle back">arrow_back</i>
            </a>
            <div v-if="!isExtension" class="session-close">
              <a @click="closeModal()">
                <i class="material-icons notranslate circle back">close</i>
              </a>
            </div>
          </div>
          <slot></slot>
        </div>
      </div>
    </div>
  </transition>
</template>

<script>
import config from "src/../config"
import { mapState, mapGetters } from "vuex"

export default {
  name: `session-frame`,
  components: {},
  props: {
    hideBack: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: ``,
    },
  },
  data: () => ({
    isExtension: config.isExtension,
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`networkSlug`]),
  },
  methods: {
    goBack() {
      try {
        this.$router.go(`-1`)
      } catch (error) {
        this.$router.push({
          name: "networks",
        })
      }
    },
    closeModal() {
      if (this.$route.meta.requiresAuth) {
        this.$router.push({
          name: "validators",
          params: {
            networkId: this.networkSlug,
          },
        })
      } else {
        // if user is signed in with address
        if (this.session.address) {
          this.$router.push({
            name: `portfolio`,
          })
          // if user is not signed in with address
        } else {
          this.$router.push({
            name: `validators`,
          })
        }
      }
    },
  },
}
</script>

<style>
@import "../../styles/session.css";

.invisible {
  visibility: hidden;
}

.component-fade-enter-active,
.component-fade-leave-active {
  transition: opacity 0.2s ease;
}

.component-fade-enter,
.component-fade-leave-to .component-fade-leave-active {
  opacity: 0;
}

.icon-circle {
  display: flex;
  justify-content: center;
  align-items: center;
  background: var(--app-fg);
  color: var(--dim);
  border-radius: 50%;
  padding: 2.5rem;
  position: absolute;
  top: 0.6rem;
  left: 40%;
  z-index: 1;
  width: 4rem;
  height: 4rem;
}

.icon-image {
  width: 2rem;
  height: 2rem;
  transform: scaleX(-1);
  filter: invert(85%) sepia(9%) saturate(18%) hue-rotate(6deg) brightness(85%)
    contrast(87%); /* converts to same than var(--dim) */
}
</style>
