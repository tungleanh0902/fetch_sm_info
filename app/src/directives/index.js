export const focusElement = {
  inserted: (el) => {
    el.focus()
  },
}

// Directive to Focus an element, but only if a child does not have focus
export const focusParentLast = {
  inserted: (el) => {
    if (!el.contains(document.activeElement)) {
      el.focus()
    }
  },
}
