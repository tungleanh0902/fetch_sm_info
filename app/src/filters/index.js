import { percent } from "../scripts/num"
import moment from "moment"
import bech32 from "bech32"

export const date = (date) => moment(date).format("MMMM Do YYYY, HH:mm")

export const fromNow = (date) => moment(date).fromNow()

export const noBlanks = function (value) {
  return value === undefined ||
    value === null ||
    value === `` ||
    value === `[do-not-modify]`
    ? `--`
    : value
}

export const percentOrPending = function (value, totalValue, pending) {
  return pending ? `--` : percent(totalValue === 0 ? 0 : value / totalValue)
}

const getAddressType = (address) => {
  if (address.startsWith("0x")) return "ethereum"
  try {
    bech32.decode(address)
    return "cosmos"
  } catch (error) {
    // ignore error
  }
  return "any"
}

export const formatAddress = (address, length = 4) => {
  if (!address) {
    return `Address Not Found`
  }
  switch (getAddressType(address)) {
    case "cosmos":
      return address.split(`1`)[0] + `…` + address.slice(-1 * length)
    case "ethereum":
      return address.slice(0, 2 + length) + `…` + address.slice(-1 * length)
    case "any":
      return address.slice(0, length) + `…` + address.slice(-1 * length)
  }
}

export const resolveValidatorName = (address, validators) => {
  // Substrate validators can have their operatorAddress as their name, which is too long and thus breaks the
  // extension size. If name is too long, then we just display the formatted address
  if (validators[address] && validators[address].name.length < 20) {
    return validators[address].name
  }
  return formatAddress(address)
}

export const validatorEntry = (validator) =>
  `${validator.name} - ${formatAddress(validator.operatorAddress, 20)}`
