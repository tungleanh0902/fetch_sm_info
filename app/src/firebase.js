import config from "../config"
import * as Sentry from "@sentry/browser"

// load firebase async
let firebaseInstance
const getFirebase = async () => {
  if (firebaseInstance) {
    return firebaseInstance
  }
  try {
    const [firebase] = await Promise.all([
      import("firebase/app"),
      import("firebase/auth"),
    ])
    firebase.initializeApp(config.firebaseConfig)
    firebaseInstance = firebase
    return firebase
  } catch (error) {
    console.error(`Firebase could not be initialized`)
    console.error(error)
    Sentry.captureException(error)
  }
}

export default getFirebase
