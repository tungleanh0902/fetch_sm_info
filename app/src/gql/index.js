/* istanbul ignore file */

import gql from "graphql-tag"
import store from "../vuex/store"

function getCurrentNetwork() {
  return store().state.connection.network
}

export const schemaMap = {
  cosmoshub: "",
  [`gaia-testnet`]: "gaia_testnet_",
  testnet: "gaia_testnet_",
}

export const ValidatorFragment = `
  id
  name
  operatorAddress
  consensusPubkey
  jailed
  picture
  details
  website
  identity
  votingPower
  startHeight
  uptimePercentage
  tokens
  commissionUpdateTime
  commission
  maxCommission
  maxChangeCommission
  status
  statusDetailed
  expectedReturns
  selfStake
`

export const AllValidators = () => {
  const currentNetwork = getCurrentNetwork()
  return gql`
    query AllValidators {
      validators(networkId: "${currentNetwork}") {
        ${ValidatorFragment}
      }
    }`
}

export const ValidatorProfile = gql`
  query validator($networkId: String!, $operatorAddress: String!) {
    validator(networkId: $networkId, operatorAddress: $operatorAddress) {
      ${ValidatorFragment}
    }
  }
`

export const DelegatorValidators = (schema) => gql`
  query ValidatorInfo($delegatorAddress: String!) {
    validators(networkId: "${schema}", delegatorAddress: $delegatorAddress) {
      ${ValidatorFragment}
    }
  }
`

export const DelegationsForDelegator = (schema) => gql`
  query delegations($delegatorAddress: String!) {
    delegations(networkId: "${schema}", delegatorAddress: $delegatorAddress) {
      id
      validator {
        ${ValidatorFragment}
      }
      amount
    }
  }
`

export const Networks = gql`
  query Networks {
    networks {
      id
      chain_id
      testnet
      title
      icon
      slug
      powered {
        name
        providerAddress
        picture
      }
    }
  }
`

// load all the data immediatly to avoid async loading later
export const NetworksAll = gql`
  query Networks($experimental: Boolean) {
    networks(experimental: $experimental) {
      id
      chain_id
      testnet
      title
      icon
      slug
      default
      lockUpPeriod
      powered {
        name
        providerAddress
        picture
      }
      feature_session
      feature_portfolio
      feature_validators
      feature_proposals
      feature_activity
      feature_explorer
      action_send
      action_claim_rewards
      action_delegate
      action_redelegate
      action_undelegate
      action_deposit
      action_vote
      action_proposal
      stakingDenom
      network_type
      address_creator
      address_prefix
      bech32_prefix
      ledger_app
      testnet
      enabled
      coinLookup {
        chainDenom
        viewDenom
        chainToViewConversionFactor
        icon
      }
      rpc_url
      HDPaths
      curves
      defaultHDPath
      defaultCurve
    }
  }
`

export const NetworksResult = (data) => data.networks

export const ProposalFragment = `
  id
  proposalId
  type
  title
  description
  creationTime
  status
  statusBeginTime
  statusEndTime
  tally {
    yes
    no
    veto
    abstain
    total
    totalVotedPercentage
  }
  deposit
  proposer {
    name
    address
    picture
  }
  validator {
    id
    name
    details
    identity
    picture
  }
  beneficiary {
    name
    address
    picture
  }
  summary
  detailedVotes {
    deposits {
      amount {
        amount
        denom
      }
      depositer {
        name
        address
        picture
      }
    }
    depositsSum
    percentageDepositsNeeded
    votes {
      id
      voter {
        name
        address
        picture
      }
      option
      amount {
        amount
        denom
      }
    }
    votesSum
    votingThresholdYes
    votingThresholdNo
    votingPercentageYes
    votingPercentageNo
    links {
      title
      link
      type
    }
    timeline {
      title 
      time
    }
  }
`

export const ProposalItem = (schema) => gql`
  query proposal($id: String!) {
    proposal(networkId: "${schema}", id: $id) {
      ${ProposalFragment}
    }
  }
`

export const GovernanceParameters = (schema) => gql`
query governanceParameters {
  governanceParameters(networkId: "${schema}") {
    depositDenom
    votingThreshold
    vetoThreshold
    depositThreshold
  }
}
`

export const Vote = (schema) => gql`
query vote($proposalId: String!, $address: String!) {
  vote(networkId: "${schema}", proposalId: $proposalId, address: $address) {
    option
  }
}
`

export const Block = (networkId) => gql`
query Block {
  block(networkId: "${networkId}") {
    id
    height
    chainId
  }
}
`

export const MetaData = (schema) => gql`
query metaData {
  metaData(networkId: "${schema}") {
    stakingDenom
  }
}
`

export const UserTransactionAdded = gql`
  subscription($networkId: String!, $address: String!) {
    userTransactionAddedV2(networkId: $networkId, address: $address) {
      hash
      height
      success
      log
    }
  }
`

export const NotificationAdded = gql`
  subscription($addressObjects: [NotificationInput]!) {
    notificationAdded(addressObjects: $addressObjects) {
      networkId
      timestamp
      title
      link
      icon
    }
  }
`

export const AddressRole = gql`
  query($networkId: String!, $address: String!) {
    accountRole(networkId: $networkId, address: $address)
  }
`
