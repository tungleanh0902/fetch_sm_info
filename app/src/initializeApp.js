/* istanbul ignore file: really just integrations */

import { Plugins } from "@capacitor/core"
import * as Sentry from "@sentry/browser"
const { App: CapacitorApp } = Plugins
import { listenToExtensionMessages } from "scripts/extension-utils"
import { checkForNewLunieVersions } from "scripts/check-for-new-lunie-versions"
import {
  enableGoogleAnalytics,
  setGoogleAnalyticsPage,
} from "scripts/google-analytics"
import config from "src/../config"
import Router, { routeGuard } from "./router"
import Store from "./vuex/store"
import { createApolloProvider, apolloErrorHandler } from "src/gql/apollo.js"
import { registerForPushNotifications } from "./scripts/pushNotifications"
import { handleDeeplink, getLaunchUrl } from "./vuex/modules/account"

if (navigator && navigator.serviceWorker) {
  // remove any existing service worker
  navigator.serviceWorker.getRegistrations().then(function (registrations) {
    for (let registration of registrations) {
      registration.unregister()
    }
  })
}

function setOptions(urlParams, store) {
  if (urlParams.experimental) {
    store.commit(`setExperimentalMode`)
  }
  if (urlParams.insecure === `true`) {
    store.commit(`setInsecureMode`)
  }
  if (urlParams.network) {
    store.dispatch(`setNetwork`, { id: urlParams.network })
  }
}

export const bootError = (error) => {
  document.querySelector("#app").innerHTML = `
  <div class="boot-error">
    <h2>Uh oh. Lunie failed to start. Our team has been notified of the issue. Please try refreshing this page in a few minutes. If the error persists please contact us at http://help.lunie.io/</h2>

    <p class="error">${error}</p>
  </div>
    `
}

export default async function init(urlParams, env = process.env) {
  let initialized = false
  // add error handlers in production
  if (env.NODE_ENV === `production`) {
    // check every minute if new Lunie versions have been deployed
    checkForNewLunieVersions()
    enableGoogleAnalytics(config.google_analytics_uid)
  }

  console.log(`Expecting backend at: ${config.graphqlHost}`)

  const apolloProvider = await createApolloProvider()
  const apolloClient = apolloProvider.clients.defaultClient

  const store = Store({ apollo: apolloClient })

  // we need to use this custom error handler as we want to use the store in there
  // we can't pass the store as it would create a circular dependency
  apolloErrorHandler.onError((error) => {
    if (error.extensions && error.extensions.code === "UNAUTHENTICATED") {
      store.dispatch("signOutUser")
      return
    }
    Sentry.captureException(error)
    console.error(error)

    if (!initialized) {
      bootError(new Error("Failed to connect to API: " + error.message))
    }
  })

  // we need to set url params before querying for networks because of experimental flag
  setOptions(urlParams, store)

  // check if user is signed in
  store.dispatch(`listenToAuthChanges`) // handles Google OAuth changes we are not influencing
  await store.dispatch(`checkSession`)

  // we load the networks first as we need them in the router
  await store.dispatch(`preloadNetworkCapabilities`)

  const router = Router(store)
  setGoogleAnalyticsPage(router.currentRoute.path)
  router.beforeEach(routeGuard(store))
  router.afterEach((to) => {
    /* istanbul ignore next */
    setGoogleAnalyticsPage(to.path)
  })

  registerForPushNotifications(store, router)

  if (config.mobileApp) {
    CapacitorApp.addListener("appUrlOpen", function (data) {
      handleDeeplink(data.url, router)
    })
    // handling deeplinks when app is opening
    getLaunchUrl(router)
  } else {
    handleDeeplink(window.location, router)
  }

  store.dispatch(`loadLocalPreferences`)
  await store.dispatch(`checkForPersistedNetwork`) // wait until signin
  store.dispatch(`checkForPersistedAddresses`)

  listenToExtensionMessages(store)

  initialized = true

  return { store, router, apolloProvider }
}
