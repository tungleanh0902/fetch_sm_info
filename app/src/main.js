/* istanbul ignore file */

import Vue from "vue"
import VTooltip from "v-tooltip"
import Vuelidate from "vuelidate"
import InfiniteScroll from "vue-infinite-scroll"
import VueClipboard from "vue-clipboard2"
import { DynamicReactiveRefs } from "vue-reactive-refs"
import { focusElement, focusParentLast } from "src/directives"
import App from "./App.vue"
import init, { bootError } from "./initializeApp"
import { getURLParams } from "scripts/url"
import { Plugins } from "@capacitor/core"
import config from "src/../config"
import * as Sentry from "@sentry/browser"
import * as Integrations from "@sentry/integrations"
import "material-design-icons-iconfont/dist/material-design-icons.css"
import AsyncComputed from "vue-async-computed"
import VueScrollTo from "vue-scrollto"

if (config.sentryDSN) {
  Sentry.init({
    dsn: config.sentryDSN,
    integrations: [new Integrations.Vue({ Vue, attachProps: true })],
    release: process.env.LUNIE_VERSION,
  })
}

Vue.config.productionTip = false

Vue.use(VTooltip)

Vue.use(Vuelidate)
Vue.use(VueClipboard)
Vue.use(InfiniteScroll)
Vue.use(AsyncComputed)
Vue.use(DynamicReactiveRefs)
Vue.use(VueScrollTo)

Vue.directive(`focus`, focusElement)
Vue.directive(`focus-last`, focusParentLast)

const urlParams = getURLParams(window)
init(urlParams)
  .then(({ store, router, apolloProvider }) => {
    const { SplashScreen, StatusBar } = Plugins

    new Vue({
      router,
      ...App,
      store,
      apolloProvider,
      mounted() {
        if (config.mobileApp) {
          SplashScreen.hide()
          StatusBar.show()
        }
      },
    }).$mount("#app")
  })
  .catch((error) => {
    Sentry.captureException(error)
    bootError(error)
  })
