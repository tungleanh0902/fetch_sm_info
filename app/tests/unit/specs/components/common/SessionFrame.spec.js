import { shallowMount, createLocalVue } from "@vue/test-utils"
import SessionFrame from "common/SessionFrame"
import { focusParentLast } from "src/directives"

let localVue = createLocalVue()
localVue.directive("focus-last", focusParentLast)

describe(`SessionFrame`, () => {
  let wrapper

  const getters = {
    networkSlug: "cosmos-hub",
  }

  beforeEach(() => {
    wrapper = shallowMount(SessionFrame, {
      localVue,
      mocks: {
        $store: { getters },
        $router: {
          go: jest.fn(),
          push: jest.fn(),
        },
        $route: {
          meta: {
            requiresAuth: true,
          },
        },
      },
      stubs: [`router-link`],
    })
  })

  it(`shows an overview of all wallets to sign in with from the extension`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should go back to Welcome when back arrow is clicked`, () => {
    wrapper.vm.goBack()
    expect(wrapper.vm.$router.go).toHaveBeenCalledWith(`-1`)
  })

  it(`should go to validators page of the current network`, () => {
    wrapper.vm.closeModal()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      name: "validators",
      params: {
        networkId: `cosmos-hub`,
      },
    })
  })
})
