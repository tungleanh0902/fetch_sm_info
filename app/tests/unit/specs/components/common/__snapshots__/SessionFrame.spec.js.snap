// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionFrame shows an overview of all wallets to sign in with from the extension 1`] = `
<div
  class="session-frame"
  mode="out-in"
  name="component-fade"
  tabindex="0"
>
  <div
    class="session-outer-container"
  >
    <!---->
     
    <div
      class="session"
    >
      <div
        class="session-header"
      >
        <a
          class=""
        >
          <i
            class="material-icons notranslate circle back"
          >
            arrow_back
          </i>
        </a>
         
        <div
          class="session-close"
        >
          <a>
            <i
              class="material-icons notranslate circle back"
            >
              close
            </i>
          </a>
        </div>
      </div>
       
    </div>
  </div>
</div>
`;
